File: windows/ChoroReader.Tests/BackIndexTests.cs
using ChoroReader.Core;

namespace ChoroReader.Tests;

/// <summary>
/// 巻末の用語索引を見つける。
///
/// <para>
/// 意味の近さで引くと索引の紙面が上位に来る。語がひたすら並んでいるので何にでも少しずつ似る。
/// <b>取りこぼしよりも誤検出のほうが痛い。</b>索引でない章を落とすと、その本文が
/// 意味検索から丸ごと消えるためである。誤検出しないことを厚めに見る。
/// </para>
/// </summary>
public class BackIndexTests
{
    // MARK: 名前の見分け

    [Theory]
    [InlineData("索引")]
    [InlineData("さくいん")]
    [InlineData("用語索引")]
    [InlineData("事項索引")]
    [InlineData("Index")]
    [InlineData("INDEX")]
    [InlineData(" 索 引 ")]
    [InlineData("索　引")]
    public void 索引の名前を見分ける(string title) => Assert.True(BackIndex.IsIndexName(title));

    /// <summary>
    /// <b>完全一致で見る。</b>部分一致にすると「索引の作り方」のような節を巻き込む。
    /// </summary>
    [Theory]
    [InlineData("索引の作り方")]
    [InlineData("第 8 章 索引")]
    [InlineData("indexing")]
    [InlineData("参考文献")]
    [InlineData("")]
    public void 索引でない名前は拾わない(string title) => Assert.False(BackIndex.IsIndexName(title));

    // MARK: EPUB

    private static EpubPublication Book(int chapters, params (string Title, int At)[] toc) => new()
    {
        Title = "検査用の本",
        Authors = [],
        ReadingOrder = [.. Enumerable.Range(0, chapters)
            .Select(at => new Link($"ch{at:00}.xhtml", "application/xhtml+xml", null,
                                   new HashSet<string>()))],
        TableOfContents = [.. toc.Select(one =>
            new TocEntry(one.Title, $"ch{one.At:00}.xhtml", null, null, []))],
        Layout = PublicationLayout.Reflowable,
        Direction = ReadingDirection.Ltr,
    };

    [Fact]
    public void 目次の索引から次の項目までを外す()
    {
        // 20 章。18 章目が索引、19 章目が奥付。
        var found = BackIndex.Of(Book(20, ("第 1 章", 0), ("索引", 18), ("奥付", 19)));

        Assert.NotNull(found);
        Assert.Equal(18, found!.Value.From);
        Assert.Equal(19, found.Value.Until);
        Assert.True(found.Value.Contains(18));
        Assert.False(found.Value.Contains(19));
    }

    /// <summary>次の項目が無ければ本の末尾まで。</summary>
    [Fact]
    public void 索引の後に何も無ければ末尾まで()
    {
        var found = BackIndex.Of(Book(20, ("第 1 章", 0), ("索引", 18)));

        Assert.Equal(new BackIndex.Span(18, 20), found);
    }

    /// <summary>
    /// <b>巻末に無ければ見ない。</b>本文の途中に「索引」という節があっても触らない
    /// （データベースの索引を説明する章など）。
    /// </summary>
    [Fact]
    public void 途中の索引という章は外さない()
    {
        // 20 章の 5 章目。0.25 の位置なので 0.8 に届かない。
        Assert.Null(BackIndex.Of(Book(20, ("第 1 章", 0), ("索引", 5), ("第 7 章", 6))));
    }

    /// <summary>同じ章を指す項目が続くことがある。<b>先へ進んだ</b>最初の項目を終わりにする。</summary>
    [Fact]
    public void 同じ章を指す項目が続いても先へ進む()
    {
        var found = BackIndex.Of(Book(20, ("索引", 18), ("あ行", 18), ("か行", 18), ("奥付", 19)));

        Assert.Equal(new BackIndex.Span(18, 19), found);
    }

    [Fact]
    public void 索引が無ければ何も外さない()
    {
        Assert.Null(BackIndex.Of(Book(20, ("第 1 章", 0), ("参考文献", 18), ("奥付", 19))));
        Assert.Null(BackIndex.Of(Book(20)));
    }

    // MARK: PDF（ページごとの見出し）

    /// <summary>ページごとの見出しを組む。<paramref name="from"/> 以降が索引。</summary>
    private static string[] Titles(int pages, int from) =>
        [.. Enumerable.Range(0, pages).Select(at => at < from ? $"第 {at / 5 + 1} 章" : "索引")];

    [Fact]
    public void 見出しが索引のページを外す()
    {
        var found = BackIndex.Of(Titles(100, 90));

        Assert.Equal(new BackIndex.Span(90, 100), found);
    }

    [Fact]
    public void 索引の後に別の節があればそこまで()
    {
        var titles = Titles(100, 90);
        titles[97] = "奥付";
        titles[98] = "奥付";
        titles[99] = "奥付";

        Assert.Equal(new BackIndex.Span(90, 97), BackIndex.Of(titles));
    }

    [Fact]
    public void 途中の索引という節は外さない()
    {
        Assert.Null(BackIndex.Of(Titles(100, 30)));
    }

    [Fact]
    public void 見出しが無ければ何も外さない()
    {
        Assert.Null(BackIndex.Of(Titles(100, 100)));
        Assert.Null(BackIndex.Of([]));
    }

    // MARK: PDF（本物の紙面）

    /// <summary>
    /// 見本の PDF には索引が無い。<b>誤検出しないこと</b>を見る。
    ///
    /// <para>
    /// 索引を持つ本での当たりは手元の蔵書で測ってあるが、実在の書籍なので git には入れない。
    /// ここで確かめられるのは、持たない本を拾わないことまでである。
    /// </para>
    /// </summary>
    [Fact]
    public void 索引を持たない見本を拾わない()
    {
        using var paper = PdfInspector.Open(Path.Combine(TestPaths.Samples, "sample.pdf"));
        var titles = BackIndex.PageTitles(paper);

        Assert.Equal(paper.PageCount, titles.Count);
        Assert.Null(BackIndex.Of(paper, titles));
    }

    /// <summary>
    /// ページごとの見出しは、次の区切りまで引き継ぐ。
    /// アウトラインのページが 0 始まりであることに乗っている（PdfOutlineTests）。
    /// </summary>
    [Fact]
    public void ページごとの見出しを引き継ぐ()
    {
        using var paper = PdfInspector.Open(Path.Combine(TestPaths.Samples, "sample.pdf"));

        // 見本は 1 ページに 1 つしおりが付いているので、そのまま並ぶ。
        Assert.Equal(["Page 1", "Page 2", "Page 3", "Page 4"], BackIndex.PageTitles(paper));
    }
}

File: macos/Tests/ChoroReaderTests/BackIndexTests.swift
import PDFKit
import XCTest
@testable import ChoroReader

/// 巻末の用語索引を見つけるところ。
///
/// **入れないと決めた紙面を、確実に外す。** 外しすぎると本文が消えるので、
/// 「拾う」より「拾いすぎない」方を厚く見る。
final class BackIndexTests: XCTestCase {
    private func entry(_ title: String, _ href: String) -> TOCEntry {
        TOCEntry(title: title, href: href)
    }

    private func publication(order: [String], toc: [TOCEntry]) -> EPUBPublication {
        EPUBPublication(title: "架空の本", authors: [], language: "ja", identifier: nil,
                        readingOrder: order.map { Link(href: $0, mediaType: "application/xhtml+xml") },
                        tableOfContents: toc, coverHref: nil,
                        layout: .reflowable, direction: .ltr)
    }

    private var chapters: [String] {
        (1 ... 10).map { "OEBPS/c\($0).xhtml" } + ["OEBPS/index.xhtml", "OEBPS/colophon.xhtml"]
    }

    // MARK: - 名前

    func test_索引の名前を見分ける() {
        for name in ["索引", "さくいん", "用語索引", "事項索引", "Index", "index", " 索 引 ", "　索引　"] {
            XCTAssertTrue(BackIndex.isIndexName(name), "「\(name)」を索引と見なせていない")
        }
    }

    /// **部分一致にしない。** 「索引の作り方」のような節を巻き込む。
    func test_索引を含むだけの見出しは索引ではない() {
        for name in ["索引の作り方", "全文索引", "索引と検索", "Indexing", "付録", ""] {
            XCTAssertFalse(BackIndex.isIndexName(name), "「\(name)」を索引と見なしている")
        }
    }

    // MARK: - EPUB

    func test_目次から索引の範囲を取る() throws {
        let made = try XCTUnwrap(BackIndex.range(in: publication(order: chapters, toc: [
            entry("第 1 章", "OEBPS/c1.xhtml"),
            entry("索引", "OEBPS/index.xhtml"),
            entry("奥付", "OEBPS/colophon.xhtml"),
        ])))
        XCTAssertEqual(made, 10 ..< 11, "索引の章だけを外していない")
    }

    /// 次の項目が無ければ本の末尾まで。
    func test_索引の後に項目が無ければ末尾まで() throws {
        let made = try XCTUnwrap(BackIndex.range(in: publication(order: chapters, toc: [
            entry("第 1 章", "OEBPS/c1.xhtml"),
            entry("索引", "OEBPS/index.xhtml"),
        ])))
        XCTAssertEqual(made, 10 ..< 12)
    }

    /// **巻末に無ければ触らない。** 本文の途中の「索引」の節は本文である。
    func test_本文の途中の索引は外さない() {
        let made = BackIndex.range(in: publication(order: chapters, toc: [
            entry("索引とは何か", "OEBPS/c2.xhtml"),
            entry("索引", "OEBPS/c3.xhtml"),
        ]))
        XCTAssertNil(made, "本文の途中を外している")
    }

    func test_索引が無ければ何も外さない() {
        let made = BackIndex.range(in: publication(order: chapters, toc: [
            entry("第 1 章", "OEBPS/c1.xhtml"),
            entry("奥付", "OEBPS/colophon.xhtml"),
        ]))
        XCTAssertNil(made)
    }

    /// 目次が空でも落ちないこと。目次を持たない本はある。
    func test_目次が無ければ何も外さない() {
        XCTAssertNil(BackIndex.range(in: publication(order: chapters, toc: [])))
    }

    // MARK: - PDF

    /// ページごとの節の見出しから、始まりと終わりが決まること。
    func test_ページの見出しから索引の範囲を取る() throws {
        var titles = [String](repeating: "第 1 章", count: 100)
        for page in 95 ..< 99 { titles[page] = "索引" }
        titles[99] = "奥付"
        XCTAssertEqual(BackIndex.range(pageTitles: titles), 95 ..< 99)
    }

    /// 最後まで索引のとき。
    func test_索引が最後まで続く() throws {
        var titles = [String](repeating: "第 1 章", count: 100)
        for page in 96 ..< 100 { titles[page] = "索引" }
        XCTAssertEqual(BackIndex.range(pageTitles: titles), 96 ..< 100)
    }

    func test_本文の途中の索引の節は外さない() {
        var titles = [String](repeating: "第 1 章", count: 100)
        for page in 20 ..< 24 { titles[page] = "索引" }
        XCTAssertNil(BackIndex.range(pageTitles: titles), "本文の途中を外している")
    }

    func test_見出しが無くても落ちない() {
        XCTAssertNil(BackIndex.range(pageTitles: []))
        XCTAssertNil(BackIndex.range(pageTitles: [String](repeating: "", count: 10)))
    }

    // MARK: - しおりの無い PDF

    /// 頁を組み立てた PDF を作る。**しおりは付けない。**
    private func pdf(_ pages: [String]) throws -> PDFKit.PDFDocument {
        let size = CGRect(x: 0, y: 0, width: 420, height: 595)
        let data = NSMutableData()
        let consumer = CGDataConsumer(data: data as CFMutableData)!
        var box = size
        let context = CGContext(consumer: consumer, mediaBox: &box, nil)!
        for text in pages {
            context.beginPDFPage(nil)
            let attributed = NSAttributedString(string: text, attributes: [
                .font: NSFont.systemFont(ofSize: 9),
            ])
            let frame = CTFramesetterCreateFrame(
                CTFramesetterCreateWithAttributedString(attributed),
                CFRangeMake(0, 0), CGPath(rect: size.insetBy(dx: 20, dy: 20), transform: nil), nil)
            CTFrameDraw(frame, context)
            context.endPDFPage()
        }
        context.closePDF()
        return try XCTUnwrap(PDFKit.PDFDocument(data: data as Data))
    }

    /// 索引らしい紙面。**行ごとに頁番号がぶら下がる。**
    private var indexPage: String {
        "索引\n" + (1 ... 40).map { "架空の項目\($0)  \($0 * 3), \($0 * 7)" }.joined(separator: "\n")
    }

    private var bodyPage: String {
        (1 ... 30).map { _ in "架空の本文である。ここには数字を置かない。" }.joined(separator: "\n")
    }

    /// **しおりが無くても、紙面から範囲が取れること。**
    func test_しおりが無くても紙面から取る() throws {
        let made = try pdf((0 ..< 18).map { _ in bodyPage } + [indexPage, indexPage])
        XCTAssertNil(BackIndex.range(pageTitles: [String](repeating: "", count: made.pageCount)),
                     "しおりから取れてしまっている（検査の前提が崩れている）")
        let found = try XCTUnwrap(BackIndex.range(in: made,
                                                  pageTitles: [String](repeating: "", count: made.pageCount)))
        XCTAssertEqual(found, 18 ..< 20)
    }

    /// **見出しが無ければ拾わない。** 数字の多い紙面（表や参考文献）を巻き込まない。
    func test_数字が多いだけの紙面は拾わない() throws {
        let table = "架空の実験結果\n" + (1 ... 40).map { "行\($0)  \($0 * 3), \($0 * 7)" }.joined(separator: "\n")
        let made = try pdf((0 ..< 18).map { _ in bodyPage } + [table, table])
        XCTAssertNil(BackIndex.range(in: made,
                                     pageTitles: [String](repeating: "", count: made.pageCount)))
    }

    /// **巻末に無ければ拾わない。** 本文の途中の索引らしい紙面は本文である。
    func test_途中の索引らしい紙面は拾わない() throws {
        let made = try pdf([indexPage] + (0 ..< 19).map { _ in bodyPage })
        XCTAssertNil(BackIndex.range(in: made,
                                     pageTitles: [String](repeating: "", count: made.pageCount)))
    }

    /// しおりがあればそちらが勝つこと。紙面を読みに行かない。
    func test_しおりがあればそれを使う() throws {
        let made = try pdf((0 ..< 18).map { _ in bodyPage } + [indexPage, indexPage])
        var titles = [String](repeating: "第 1 章", count: made.pageCount)
        titles[19] = "索引"
        XCTAssertEqual(BackIndex.range(in: made, pageTitles: titles), 19 ..< 20,
                       "しおりより紙面を優先している")
    }
}
